test-data/unit/check-selftype.test

```diff
@@ -1267,3 +1267,206 @@ class Test(Generic[T]):
+reveal_type(bad)  # N: Revealed type is "def (Any) -> Any"
+
+def func() -> None:
+    var: Self  # E: Self type is only allowed in annotations within class definition
+
+class C(Self): ...  # E: Self type is only allowed in annotations within class definition
+Alias = Callable[[Self], Self]  # E: Self type is only allowed in annotations within class definition
+a: Alias
+reveal_type(a)  # N: Revealed type is "def (Any) -> Any"
+
+[case testTypingSelfConflict]
+from typing import Self, TypeVar, Tuple
+
+T = TypeVar("T")
+class C:
+    def meth(self: T) -> Tuple[Self, T]: ...  # E: Method cannot have explicit self annotation and Self type
+reveal_type(C().meth())  # N: Revealed type is "Tuple[<nothing>, __main__.C]"
+[builtins fixtures/property.pyi]
+
+[case testTypingSelfProperty]
+from typing import Self, Tuple
+class C:
+    @property
+    def attr(self) -> Tuple[Self, ...]: ...
+class D(C): ...
+
+reveal_type(D().attr)  # N: Revealed type is "builtins.tuple[__main__.D, ...]"
+[builtins fixtures/property.pyi]
+
+[case testTypingSelfCallableVar]
+from typing import Self, Callable
+
+class C:
+    x: Callable[[Self], Self]
+    def meth(self) -> Callable[[Self], Self]: ...
+class D(C): ...
+
+reveal_type(C().x)  # N: Revealed type is "def (__main__.C) -> __main__.C"
+reveal_type(D().x)  # N: Revealed type is "def (__main__.D) -> __main__.D"
+reveal_type(D().meth())  # N: Revealed type is "def (__main__.D) -> __main__.D"
+
+[case testTypingSelfClassMethod]
+from typing import Self
+
+class C:
+    @classmethod
+    def meth(cls) -> Self: ...
+    @staticmethod
+    def bad() -> Self: ...  # E: Static methods cannot use Self type \
+        # E: A function returning TypeVar should receive at least one argument containing the same TypeVar \
+        # N: Consider using the upper bound "C" instead
+
+class D(C): ...
+reveal_type(D.meth())  # N: Revealed type is "__main__.D"
+reveal_type(D.bad())  # N: Revealed type is "<nothing>"
+[builtins fixtures/classmethod.pyi]
+
+[case testTypingSelfOverload]
+from typing import Self, overload, Union
+
+class C:
+    @overload
+    def foo(self, other: Self) -> Self: ...
+    @overload
+    def foo(self, other: int) -> int: ...
+    def foo(self, other: Union[Self, int]) -> Union[Self, int]:
+        return other
+class D(C): ...
+reveal_type(D().foo)  # N: Revealed type is "Overload(def (other: __main__.D) -> __main__.D, def (other: builtins.int) -> builtins.int)"
+
+[case testTypingSelfNestedInAlias]
+from typing import Generic, Self, TypeVar, List, Tuple
+
+T = TypeVar("T")
+Pairs = List[Tuple[T, T]]
+
+class C(Generic[T]):
+    def pairs(self) -> Pairs[Self]: ...
+class D(C[T]): ...
+reveal_type(D[int]().pairs())  # N: Revealed type is "builtins.list[Tuple[__main__.D[builtins.int], __main__.D[builtins.int]]]"
+[builtins fixtures/tuple.pyi]
+
+[case testTypingSelfOverrideVar]
+from typing import Self, TypeVar, Generic
+
+T = TypeVar("T")
+class C(Generic[T]):
+    x: Self
+
+class D(C[int]):
+    x: D
```

Comment:
What if we now have a class like this:
```py
class E(D): pass
```
This should perhaps be an error (incompatible attribute), since the type of `x` is `D` in `E`, which could be considered incompatible with the definition in `C`. It looks like we don't have a similar check for methods, so it's okay to not generate an error.

Reply:
Yeah, this looks wrong, but also may be not super easy to implement. Let's postpone this for now.

Reply:
Worth adding an xfail test?

Reply:
TBH it is unlikely it will be fixed accidentally, instead I will open a follow up issue, maybe at some point we will have energy to tighten this.